feat(rome_js_analyze): promote rules

ematipico opened the pull request:
<!--
	Thanks for submitting a pull request!

	We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request.

	Once created, your PR will be automatically labeled according to changed files.

	Learn more about contributing: https://github.com/rome/tools/blob/main/CONTRIBUTING.md
-->

## Summary

Closes #3503 

- added support for spread props in `useAltText`
- updated outdated snapshots inside the CLI
- updated configuration files inside the snapshots of the CLI
- created a new group called `security` as per https://github.com/rome/tools/pull/3471#issuecomment-1290471908
- created a new group called `complexity`as per https://github.com/rome/tools/discussions/3461
- the rule `useCamelCase` stays in `nursery` because it's still full of false positives (lots of them happen in our repo)
- moves existing rules in more suited groups
- I decided to recommend `noUnreachable` because we need data in order to make the CFG better, and this rule is perfect because because it reaches the boundaries of the feature. Although it means that we will need to focus out attention on possible bugs: @leops @xunilrj , what do you think? 


This PR promotes the following rules:
| Rule name                               | Old Group     | New group     | Reason                      | Recommended |
|-----------------------------------------|---------------|---------------|-----------------------------|-------------|
| `noPositiveTabindex`                    | `nursery`     | `a11y`        | It focuses on accessibility | `true`      |
| `useKeyWithMouseEvents`                 | `nursery`     | `a11y`        | It focuses on accessibility | `true`      |
| `useAnchorContent`                      | `nursery`     | `a11y`        | It focuses on accessibility | `true`      |
| `useBlankTarget`                        | `nursery`     | `a11y`        | It focuses on accessibility | `true`      |
| `useValidAnchor`                        | `nursery`     | `a11y`        | It focuses on accessibility | `true`      |
| `useKeyWithClickEvents`                 | `nursery`     | `a11y`        | It focuses on accessibility | `true`      |
| `useButtonType`                         | `nursery`     | `a11y`        | It focuses on accessibility | `true`      |
| `noAutofocus`                           | `nursery`     | `a11y`        | It focuses on accessibility | `true`      |
| `useFragmentSyntax`                     | `nursery`     | `style`       | It's just syntax sugar      | `true`      |
| `useTemplate`                           | `correctness` | `style`       | Just a styling decision     | `true`      |
| `useBlockStatements`                    | `correctness` | `style`       | Just a styling decision     | `true`      |
| `noImplicitBoolean`                     | `correctness` | `style`       | Just a styling decision     | `true`      |
| `useCamelCase`                          | `nursery`     | `style`       | Just a styling decision     | `true`      |
| `useOptionalOperator`                   | `nursery`     | `style`       | Syntax super for checks     | `true`      |
| `noVoidElementsWithChildren`            | `nursery`     | `correctness` | They help to prevent bugs   | `true`      |
| `noChildrenProp`                        | `nursery`     | `correctness` | They help to prevent bugs   | `true`      |
| `noArrayIndexKey`                       | `nursery`     | `correctness` | They help to prevent bugs   | `true`      |
| `noRenderReturnValue`                   | `nursery`     | `correctness` | They help to prevent bugs   | `true`      |
| `noNewSymbol`                           | `nursery`     | `correctness` | They help to prevent bugs   | `true`      |
| `noUselessFragments`                    | `nursery`     | `correctness` | They help to prevent bugs   | `true`      |
| `noUnusedVariables`                     | `nursery`     | `correctness` | They help to prevent bugs   | `false`      |
| `noUnreachable`                         | `nursery`     | `correctness` | They help to prevent bugs   | `true`      |
| `noUnusedVariables`                     | `nursery`     | `correctness` | They help to prevent bugs   | `true`      |
| `noRestrictedGlobals`                   | `nursery`     | `correctness` | They help to prevent bugs   | `true`      |
| `useSimplifiedLogicExpression`          | `correctness` | `complexity`  | Prevents complexity         | `true`      |
| `noExtraBooleanCast`                    | `correctness` | `complexity`  | Prevents complexity         | `true`      |
| `noDangerouslySetInnerHtml`             | `nursery`     | `security`    | Needs extra attention       | `true`      |
| `noDangerouslySetInnerHtmlWithChildren` | `nursery`     | `security`    | Needs extra attention       | `true`      |
| `noDebugger`                            | `correctness` | `security`    | Needs extra attention       | `true`      |


<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

<!-- Link any relevant issues if necessary or include a transcript of any Discord discussion. -->

## Test Plan

Snapshot tests should only have their category updated. If there's some code frame changed/removed, it means something has been incorrectly moved.

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output. -->

MichaReiser commented:
I think we should avoid framework-specific group naming for the same reason we avoided language-specific group naming. Our configuration should have a `framework` section instead where it is possible to configure the used frameworks and the analyzer then enables/disabled rules based on that configuration. 

For example: `noDangerouslySetInnerHtml` should be in the `security` group and can be generalized to `noHtmlInjection`. 

I do understand that we don't have the infrastructure to accomplish this today but that means we should prioritise adding the necessary capabilities to Rome because renaming rules is a confusing experience for our users.

ematipico commented:
> I think we should avoid framework-specific group naming for the same reason we avoided language-specific group naming.

How would you solve a possible issue where a rule that targets React goes in conflict with another library that uses JSX?

MichaReiser commented:
> > I think we should avoid framework-specific group naming for the same reason we avoided language-specific group naming.
> 
> How would you solve a possible issue where a rule that targets React goes in conflict with another library that uses JSX?

Our configuration should have a framework section instead where it is possible to configure the used frameworks. The analyzer then enables/disabled rules based on that configuration. For example. The react rules only get enabled if you configure the react framework (with an optional version) or, as an alternative, by setting a `javascript.jsxDialect` to `react`. 

The benefit of this approach is that transformation (and maybe even optimisations and bundling) must know about the used frameworks because the transformation may slightly differ.

Boshen commented:
I'll help out and test these promoted rules tomorrow (using internal code @ ByteDance).

ematipico commented:
@leops  and @xunilrj , could you please chime in and let us know what you think?

MichaReiser commented:
On the grouping: 

* noUselessFragments: How is this preventing bugs? In my view, this is more about writing idiomatic JSX and should thus be in the `style` group (or maybe `complexity`)
* useOptionalOperator: Same as for `noUselessFragements`. It isn't preventing bugs but helps you write more idiomatic JavaScript. I could otherwise see this being a `complexity` rule but I would favour `style` because using `?.` doesn't reduce complexity. Which is the case for clippys `?` [operator rule](https://rust-lang.github.io/rust-clippy/stable/index.html#question_mark)




https://github.com/rust-lang/rust-clippy

ematipico commented:
> useOptionalOperator: Same as for noUselessFragements. It isn't preventing bugs but helps you write more idiomatic JavaScript. I could otherwise see this being a complexity rule but I would favour style because using ?. doesn't reduce complexity. Which is the case for clippys ? [operator rule](https://rust-lang.github.io/rust-clippy/stable/index.html#question_mark)

I was on the fence on this one. Initially I thought about `style` but then thought that styling issue could be easily overlooked by the users because it cold be seen as "an opinionated set of rules", which a user might just turn off by default. But then I thought about my previous experiences where I encountered `cannot read of undefined` because of a missed a check (let's leave TS out of the equation). Having it in a `style` group feels like not giving enough importance... **but technically, it's a `style` rule**, ultimately.

MichaReiser commented on xtask/lintdoc/src/main.rs:
```diff
@@ -88,17 +88,100 @@ fn main() -> Result<()> {
+            "correctness" => (
+                "Correctness",
+                markup! {
+                    "This group should includes those rules that are meant to prevent possible bugs and misuse of the language."
```
You have a merge issue here. It reverts my changes to have the `nurserey_rules` last and also reverts some of the wording changes that I made to `correctness`, `style`, and `nurserey`

MichaReiser commented on xtask/lintdoc/src/main.rs:
```diff
@@ -88,17 +88,100 @@ fn main() -> Result<()> {
+            ),
+            "security" => (
+                "Security",
+                markup! {
+                    "Rules that require special attention. If if the code can be correct and bug-free, publishing it can be an hazard."
+                },
```
Rules that detect potential security flaws.

MichaReiser commented on xtask/lintdoc/src/main.rs:
```diff
@@ -88,17 +88,100 @@ fn main() -> Result<()> {
+            ) {
+                Ok(summary) => {
+                    writeln!(index, "<div class=\"rule\">")?;
+                    writeln!(index, "<h3 data-toc-exclude id=\"{rule}\">")?;
```
The code here also undoes my changes that ensure that lint rules show up in the TOC

MichaReiser reviewed (changes requested):
Thanks for working out all the new groups. There's a merge issue in the codegen that must be resolved and I do have a question around the grouping of `noDebugger`:
What are the security concerns around `debugger`? I think if there's a security concern then the lint diagnostic should point that out. I lean towards `correctness` because its useless code in production or has the undesired effect of stopping execution all together

ematipico commented:
> Thanks for working out all the new groups. There's a merge issue in the codegen that must be resolved and I do have a question around the grouping of `noDebugger`:
> What are the security concerns around `debugger`? I think if there's a security concern then the lint diagnostic should point that out. I lean towards `correctness` because its useless code in production or has the undesired effect of stopping execution all together


My concern is around the fact that deploying code with debugger, would inevitably trigger a breakpoint if someone opens the developer console or runs the debugger in a backend app. Imagine being a library author and a `debugger` gets published.

I don't consider the code useless because. Maybe a misuse of the language but it's also debatable.

MichaReiser commented:
> My concern is around the fact that deploying code with debugger, would inevitably trigger a breakpoint if someone opens the developer console or runs the debugger in a backend app. Imagine being a library author and a debugger gets published.

That's why I would consider this a bug/broken code. It is code with unintended effects. I share your sentiment that it doesn't perfectly falls into the `correctness` group but I don't see a security concern.

ematipico commented:
@MichaReiser merge conflicts fixed and `noDebugger` moved under `correctness`

MichaReiser reviewed (approved):
Looking good from a code perspective.

Should we re-consider promoting `noUnusedVariables` considering that it still reports a fast number of false positives? 

Do you plan to test the new rules on real projects before merging to verify that they meet our quality criteria?

ematipico commented:
> Looking good from a code perspective.
> 
> Should we reconsider promoting `noUnusedVariables` considering that it still reports a vast number of false positives?
> 

The current bugs have been closed. For now I moved it to a stable group and marked as *not recommended*. 

> Do you plan to test the new rules on real projects before merging to verify that they meet our quality criteria?

Yeah I did, and fixed some false positives in other PRs. We should be good to go.
